include/blmc_robots/n_joint_blmc_robot_driver.hpp

```diff
@@ -190,15 +193,73 @@ class NJointBlmcRobotDriver
         double start_time_sec = real_time_tools::Timer::get_current_time_sec();
 
         Observation observation = get_latest_observation();
-        Action applied_action =
-            mct::clamp(desired_action, -max_torque_Nm_, max_torque_Nm_);
-        applied_action =
-            applied_action - safety_kd_.cwiseProduct(observation.velocity);
-        applied_action =
-            mct::clamp(applied_action, -max_torque_Nm_, max_torque_Nm_);
-
-        joint_modules_.set_torques(applied_action);
+
+        Action applied_action;
+
+        applied_action.torque = desired_action.torque;
+        applied_action.position = desired_action.position;
+
+        // Position controller
+        // -------------------
+        // TODO: add position limits
+        //
+        // NOTE: in the following lines, the Eigen function `unaryExpr` is
+        // used to determine which elements of the vectors are NaN.  This is
+        // because `isNaN()` was only added in Eigen 3.3 but we have to be
+        // compatible with 3.2.
+        // The std::isnan call needs to be wrapped in a lambda because it is
+        // overloaded and unaryExpr cannot resolve which version to use when
+        // passed directly.
+
+        // Run the position controller only if a target position is set for at
+        // least one joint.
+        if (!applied_action.position
+                 .unaryExpr([](double x) { return std::isnan(x); })
+                 .all())
+        {
+            // Replace NaN-values with default gains
+            applied_action.position_kp =
+                applied_action.position_kp
+                    .unaryExpr([](double x) { return std::isnan(x); })
+                    .select(default_position_kp_, desired_action.position_kp);
+            applied_action.position_kd =
+                applied_action.position_kd
+                    .unaryExpr([](double x) { return std::isnan(x); })
+                    .select(default_position_kd_, desired_action.position_kd);
+
+            Vector position_error = applied_action.position - observation.angle;
```

Comment:
i guess we should now also change it from angle to position in the Observation class.

Reply:
Yes, I will do this next